### PR TITLE
Initial commit of metrics for scaling

### PR DESCRIPTION
**Pull request by dklyle**
The beginning for #117

### CODE REVIEW

File: metrics/scaling/k8s_scale.sh [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+
+# $1 is the launch time in seconds this pod/container took to start up.
+# $2 is the number of pod/containers under test
+grab_stats() {
+	local launch_time=$1
+	local n_pods=$2
+	info "And grab some stats"
+	# Tell mpstat to measure over a short period, not only so we get slightly de-noised data, but also
+	# if you don't tell it the period, you will get the avg since boot, which is not what we want.
+	local cpu_idle=$(kubectl exec -ti ${stats_pod} -- sh -c "mpstat -u 3 1 | tail -1 | awk '{print \$11}'" | sed 's/\r//')
```
Comment (grahamwhaley) [inline metrics/scaling/k8s_scale.sh]: heh, I'd forgotten I wrote that! ;-)

File: metrics/scaling/k8s_scale.sh [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+		    -e "s|@LABEL@|${LABEL}|g" \
+		    -e "s|@LABELVALUE@|${LABELVALUE}|g" \
+		    < ${input_yaml} > ${generated_yaml}
+    fi
+
+		info "Applying changes"
+		# FIXME - time taken in 'seconds' - we can do better than that, if we
+		# use the date nanosecond stamp, and steal the magic from the density test script.
+        # switch to 'date +%N'
+		local start_time=$(date +%s)
```
Comment (grahamwhaley) [inline metrics/scaling/k8s_scale.sh]: the json.sh lib file has ms and ns time functions we can now use here I reckon..

File: metrics/scaling/k8s_scale.sh [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+	# We could check we have just the one node here - right now this is a single node
+	# test!! - because, our stats gathering is rudimentry, as k8s does not provide
+	# a nice way to do it (unless you want to parse 'descibe nodes')
+	# Have a read of https://github.com/kubernetes/kubernetes/issues/25353
+
+	# FIXME - check the node(s) can run enough pods - check 'max-pods' in the
+	# kubelet config - from 'kubectl describe node -o json' ?
+
+	# Launch our stats gathering pod
+	kubectl apply -f ${stats_pod}.yaml
```
Comment (grahamwhaley) [inline metrics/scaling/k8s_scale.sh]: I've just found, we might want to prepend the `${SCRIPT_PATH}` to the front of the yaml files etc., as if you don't run the test from within its own directory right now it fails to find `stats.yaml` for instance.

Comment (dklyle) [reply metrics/scaling/k8s_scale.sh]: that will be true for bb.yaml.in too, it's mergeable without updating, but will fix

Comment (dklyle) [reply metrics/scaling/k8s_scale.sh]: updated for bb.yaml.in, generated.yaml, and stats.yaml

File: metrics/scaling/k8s_scale.sh (hunk 2 above)
Comment (dklyle) [reply metrics/scaling/k8s_scale.sh]: changed to milliseconds

File: metrics/scaling/bb.yaml.in [hunk 4]
```diff
@@ -0,0 +1,33 @@
+apiVersion: apps/v1
```
[the remark this answers is not in the document]
Comment (dklyle) [reply metrics/scaling/bb.yaml.in]: added

File: metrics/lib/kata-common.bash [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+# Place where virtcontainers keeps its active pod info
+VC_POD_DIR="${VC_POD_DIR:-/var/lib/vc/sbs}"
+
+# Sandbox runtime directory
+RUN_SBS_DIR="${RUN_SBS_DIR:-/run/vc/sbs}"
+
+KATA_HYPERVISOR="${KATA_HYPERVISOR:-qemu}"
+
+die() {
```
[the remark this answers is not in the document]
Comment (dklyle) [reply metrics/lib/kata-common.bash]: removed from common.bash as in sources kata-common.bash

File: metrics/lib/json.bash [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	if [ "$1" == "k8s" ]; then
+		# FIXME - add k8s specific data dump here.
+		true
+	else
+		local iskata=$(is_a_kata_runtime "$RUNTIME")
+		if [ "$iskata" == "1" ]; then
+			local rpath="$(get_docker_kata_path $RUNTIME)"
+			local json="$(cat << EOF
+	"kata-env" :
+	$($rpath kata-env --json)
```
[the remark this answers is not in the document]
Comment (dklyle) [reply metrics/lib/json.bash]: punted on this for now

File: metrics/lib/common.bash [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+LIB_DIR=${THIS_FILE%/*}
+RESULT_DIR="${LIB_DIR}/../results"
+
+source ${LIB_DIR}/kata-common.bash
+source ${LIB_DIR}/json.bash
+source /etc/os-release || source /usr/lib/os-release
+KATA_KSM_THROTTLER="${KATA_KSM_THROTTLER:-no}"
+
+# Set variables to reasonable defaults if unset or empty
+DOCKER_EXE="${DOCKER_EXE:-docker}"
```
[the remark this answers is not in the document]
Comment (dklyle) [reply metrics/lib/common.bash]: punted on this for now, will work on once this is merged

File: metrics/lib/common.bash [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+THIS_FILE=$(readlink -f ${BASH_SOURCE[0]})
+LIB_DIR=${THIS_FILE%/*}
+RESULT_DIR="${LIB_DIR}/../results"
+
+source ${LIB_DIR}/kata-common.bash
+source ${LIB_DIR}/json.bash
+source /etc/os-release || source /usr/lib/os-release
+KATA_KSM_THROTTLER="${KATA_KSM_THROTTLER:-no}"
```
[the remark this answers is not in the document]
Comment (dklyle) [reply metrics/lib/common.bash]: removed and the associated function

File: metrics/lib/common.bash [hunk 9]
```diff
@@ -0,0 +1,339 @@
+#!/bin/bash
+#
+# Copyright (c) 2017,2018 Intel Corporation
```
[the remark this answers is not in the document]
Comment (dklyle) [reply metrics/lib/common.bash]: updated

File: metrics/scaling/k8s_scale.sh [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+	info "Running test"
+
+	trap cleanup EXIT QUIT KILL
+
+	metrics_json_start_array
+	for reqs in $(seq ${STEP} ${STEP} ${NUM_PODS}); do
+		info "Testing replicas ${reqs} of ${NUM_PODS}"
+		# Generate the next yaml file
+
+	local runtime_command
```
Comment (grahamwhaley) [inline metrics/scaling/k8s_scale.sh]: just noticed, I think indentation is bust in this section - it is all in the `for` loop, so some of the lines need to be pushed in one tab.

Comment (krsna1729) [reply metrics/scaling/k8s_scale.sh]: we should incorporate these, maybe as git pre-commit hooks?
https://github.com/mvdan/sh
https://github.com/koalaman/shellcheck

Comment (grahamwhaley) [reply metrics/scaling/k8s_scale.sh]: @krsna1729 eek. maybe.... I think we opted out of `shellcheck` over in kata land as there were some things it moaned about that were particularly hard to get around iirc...

Let's see if anybody finds the time to try these out on the code.... and, reserve the right to not inflict them if they pain barrier will be too high ;-)

**Comment by grahamwhaley** [comment]
Merging - we will update and clean some other bits in other PRs as we grow and improve the code base.

**Pull request merged**